source/simulator/solver_schemes.cc

```diff
@@ -1299,9 +1143,35 @@ namespace aspect
                       (1.0-(dcr.newton_residual_for_derivative_scaling_factor/dcr.switch_initial_residual))));
 
         do_one_defect_correction_Stokes_step(dcr, use_picard);
-        relative_residual = dcr.residual/dcr.initial_residual;
 
-        pcout << std::endl;
+        double max = 0.0;
+        for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+          {
+            // in models with melt migration the melt advection equation includes the divergence of the velocity
+            // and can not be expected to converge to a smaller value than the residual of the Stokes equation.
+            // thus, we set a threshold for the initial composition residual.
+            // this only plays a role if the right-hand side of the advection equation is very small.
+            const double threshold = (parameters.include_melt_transport && c == introspection.compositional_index_for_name("porosity")
+                                      ?
+                                      parameters.linear_stokes_solver_tolerance * time_step
+                                      :
+                                      0.0);
+            if (initial_composition_residual[c]>threshold)
+              max = std::max(relative_composition_residual[c],max);
+          }
+
+        // write the residual output in the same order as the solutions
+        pcout << "      Relative nonlinear residuals (temperature, compositional fields, Stokes system): " << relative_temperature_residual;
```

Comment:
Sorry to dump this on you, but because we are modifying this output anyway: Can you change this line into the following, here and in the `solve_iterated_advection_and_stokes()` function:

```suggestion
        pcout << "      Relative nonlinear residuals (temperature"
              << (introspection.n_compositional_fields > 0 ? ", compositional fields" : "")
              << ", Stokes system):"
              << relative_temperature_residual;
```

This way the output should only show the compositional fields if there are any.